### PR TITLE
Model changes to add application_instance_id

### PR DESCRIPTION
**Pull request by marcospri**
Model changes for the migration in https://github.com/hypothesis/lms/pull/3748. 

Needs to the deployed after the migration is deployed.

### CODE REVIEW

File: tests/unit/lms/services/group_info_test.py [hunk 1]
```diff
@@ -119,7 +119,7 @@ def params(self):
         return {
             column: f"TEST_{column.upper()}"
             for column in GroupInfo.columns()
-            if column not in ("consumer_key", "_info")
+            if column not in ("consumer_key", "_info", "application_instance_id")
```
Comment (marcospri) [inline tests/unit/lms/services/group_info_test.py:122]: This automatic param generation based on the name only it's not great. 

Noted to revisit the service and test for a refactor when this settles.

File: lms/models/application_instance.py [hunk 2]
```diff
@@ -87,15 +87,19 @@ class Product(str, Enum):
     #: (each token belongs to a different user of this application
     #: instance's LMS).
     access_tokens = sa.orm.relationship(
-        "OAuth2Token", back_populates="application_instance"
+        "OAuth2Token",
+        back_populates="application_instance",
+        foreign_keys="OAuth2Token.consumer_key",
```
Comment (marcospri) [inline lms/models/application_instance.py:92]: This `foreign_keys` is needed on the SQLA side to know what FK should the relationship follow in the backwards relationship to access_tokens and group_infos now that there are two FK pointing to `application_instances`

File: lms/models/group_info.py [hunk 3]
```diff
@@ -40,9 +40,16 @@ class GroupInfo(BASE):
...
+    application_instance_id = sa.Column(
+        sa.Integer(),
+        sa.ForeignKey("application_instances.id", ondelete="cascade"),
+        nullable=True,
+    )
+
     #: The ApplicationInstance that this group belongs to.
     application_instance = sa.orm.relationship(
-        "ApplicationInstance", back_populates="group_infos"
+        "ApplicationInstance", back_populates="group_infos", foreign_keys=[consumer_key]
```
Comment (marcospri) [inline lms/models/group_info.py:52]: This is the other side of the relationship, `foreign_keys` also needed here to know which FK to follow.

In this PR, as application_instance_id is null-able everything still uses `consumer_key`

File: lms/models/application_instance.py [hunk 4]
```diff
@@ -87,15 +87,19 @@ class Product(str, Enum):
...
 
     #: A list of all the courses for this application instance.
     courses = sa.orm.relationship("LegacyCourse", back_populates="application_instance")
 
     #: A list of all the GroupInfo's for this application instance.
     group_infos = sa.orm.relationship(
-        "GroupInfo", back_populates="application_instance"
+        "GroupInfo",
+        back_populates="application_instance",
+        foreign_keys="GroupInfo.consumer_key",
```
Comment (marcospri) [inline lms/models/application_instance.py:102]: Same, we point application_instances.group_infos to make the queries to use GroupInfo.consumer_key.

File: lms/models/grading_info.py [hunk 5]
```diff
@@ -48,6 +48,14 @@ class GradingInfo(CreatedUpdatedMixin, BASE):
     lis_result_sourcedid = sa.Column(sa.UnicodeText(), nullable=False)
     lis_outcome_service_url = sa.Column(sa.UnicodeText(), nullable=False)
     oauth_consumer_key = sa.Column(sa.UnicodeText(), nullable=False)
+
+    #: The ApplicationInstance that this group belongs to foreign key
```
Comment (jon-betts) [inline lms/models/grading_info.py:52]: What's up with this weird comment style in all these files?

File: tests/unit/lms/services/group_info_test.py (hunk 1 above)
Comment (jon-betts) [reply tests/unit/lms/services/group_info_test.py:122]: Eh. :shrug: Maybe this should just be hard-coded?

**Pull request merged**